core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala

```diff
@@ -1281,6 +1286,59 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with BeforeAndAfterE
     assert(master.getLocations("item").isEmpty)
   }
 
+  test("SPARK-20640: Shuffle registration timeout and maxAttempts conf are working") {
+    val tryAgainMsg = "test_spark_20640_try_again"
+    // a server which delays response 50ms and must try twice for success.
+    def newShuffleServer(port: Int): (TransportServer, Int) = {
+      val attempts = new mutable.HashMap[String, Int]()
+      val handler = new NoOpRpcHandler {
+        override def receive(client: TransportClient, message: ByteBuffer,
```

Comment:
Updated. By the way, I am a little confused.

First, when you insert line break before, Intellij auto indent like this:

```
override def receive(
                                  client: TransportClient
```

Second, in the same file, at near 1349, `fetchBlocks`'s indent is like this:

```
    override def fetchBlocks(
        host: String,
        port: Int,
        execId: String,
        blockIds: Array[String],
        listener: BlockFetchingListener,
        shuffleFiles: Array[File]): Unit = {
```